From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Natalie Vock <natalie.vock@gmx.de>
Date: Fri, 16 Aug 2024 09:55:25 +0200
Subject: [PATCH 07/17] STEAMOS: radv: Pass vk_sync_waits through to
 radv_bo_virtual_bind

So we can later pass the syncobjs through to the kernel.

(cherry picked from commit 4bb920c441cc3c3eccbe37572cc1f8e28331b903)
---
 src/amd/vulkan/radv_buffer.c |  3 ++-
 src/amd/vulkan/radv_buffer.h |  3 ++-
 src/amd/vulkan/radv_queue.c  | 29 ++++++++++++++++++-----------
 3 files changed, 22 insertions(+), 13 deletions(-)

diff --git a/src/amd/vulkan/radv_buffer.c b/src/amd/vulkan/radv_buffer.c
index 645654f153c..651dae21acc 100644
--- a/src/amd/vulkan/radv_buffer.c
+++ b/src/amd/vulkan/radv_buffer.c
@@ -308,7 +308,8 @@ radv_bo_destroy(struct radv_device *device, struct vk_object_base *object, struc
 
 VkResult
 radv_bo_virtual_bind(struct radv_device *device, struct vk_object_base *object, struct radeon_winsys_bo *parent,
-                     uint64_t offset, uint64_t size, struct radeon_winsys_bo *bo, uint64_t bo_offset)
+                     uint64_t offset, uint64_t size, struct radeon_winsys_bo *bo, uint64_t bo_offset, uint32_t wait_count,
+                     const struct vk_sync_wait *waits)
 {
    struct radv_physical_device *pdev = radv_device_physical(device);
    struct radv_instance *instance = radv_physical_device_instance(pdev);
diff --git a/src/amd/vulkan/radv_buffer.h b/src/amd/vulkan/radv_buffer.h
index 7e194eee230..072a0aab968 100644
--- a/src/amd/vulkan/radv_buffer.h
+++ b/src/amd/vulkan/radv_buffer.h
@@ -37,7 +37,8 @@ VkResult radv_bo_create(struct radv_device *device, struct vk_object_base *objec
 
 VkResult radv_bo_virtual_bind(struct radv_device *device, struct vk_object_base *object,
                               struct radeon_winsys_bo *parent, uint64_t offset, uint64_t size,
-                              struct radeon_winsys_bo *bo, uint64_t bo_offset);
+                              struct radeon_winsys_bo *bo, uint64_t bo_offset, uint32_t wait_count,
+                              const struct vk_sync_wait *waits);
 
 void radv_bo_destroy(struct radv_device *device, struct vk_object_base *object, struct radeon_winsys_bo *bo);
 
diff --git a/src/amd/vulkan/radv_queue.c b/src/amd/vulkan/radv_queue.c
index 7d4751fdc96..b1ad0772e4c 100644
--- a/src/amd/vulkan/radv_queue.c
+++ b/src/amd/vulkan/radv_queue.c
@@ -36,7 +36,8 @@ radv_get_queue_global_priority(const VkDeviceQueueGlobalPriorityCreateInfo *pObj
 }
 
 static VkResult
-radv_sparse_buffer_bind_memory(struct radv_device *device, const VkSparseBufferMemoryBindInfo *bind)
+radv_sparse_buffer_bind_memory(struct radv_device *device, const VkSparseBufferMemoryBindInfo *bind, uint32_t wait_count,
+                               const struct vk_sync_wait *waits)
 {
    VK_FROM_HANDLE(radv_buffer, buffer, bind->buffer);
    VkResult result = VK_SUCCESS;
@@ -66,7 +67,7 @@ radv_sparse_buffer_bind_memory(struct radv_device *device, const VkSparseBufferM
       }
       if (size) {
          result = radv_bo_virtual_bind(device, &buffer->vk.base, buffer->bo, resourceOffset, size, mem ? mem->bo : NULL,
-                                       memoryOffset);
+                                       memoryOffset, wait_count, waits);
          if (result != VK_SUCCESS)
             return result;
       }
@@ -77,14 +78,15 @@ radv_sparse_buffer_bind_memory(struct radv_device *device, const VkSparseBufferM
    }
    if (size) {
       result = radv_bo_virtual_bind(device, &buffer->vk.base, buffer->bo, resourceOffset, size, mem ? mem->bo : NULL,
-                                    memoryOffset);
+                                    memoryOffset, wait_count, waits);
    }
 
    return result;
 }
 
 static VkResult
-radv_sparse_image_opaque_bind_memory(struct radv_device *device, const VkSparseImageOpaqueMemoryBindInfo *bind)
+radv_sparse_image_opaque_bind_memory(struct radv_device *device, const VkSparseImageOpaqueMemoryBindInfo *bind, uint32_t wait_count,
+                                     const struct vk_sync_wait *waits)
 {
    VK_FROM_HANDLE(radv_image, image, bind->image);
    VkResult result;
@@ -96,7 +98,7 @@ radv_sparse_image_opaque_bind_memory(struct radv_device *device, const VkSparseI
          mem = radv_device_memory_from_handle(bind->pBinds[i].memory);
 
       result = radv_bo_virtual_bind(device, &image->vk.base, image->bindings[0].bo, bind->pBinds[i].resourceOffset,
-                                    bind->pBinds[i].size, mem ? mem->bo : NULL, bind->pBinds[i].memoryOffset);
+                                    bind->pBinds[i].size, mem ? mem->bo : NULL, bind->pBinds[i].memoryOffset, wait_count, waits);
       if (result != VK_SUCCESS)
          return result;
    }
@@ -105,7 +107,8 @@ radv_sparse_image_opaque_bind_memory(struct radv_device *device, const VkSparseI
 }
 
 static VkResult
-radv_sparse_image_bind_memory(struct radv_device *device, const VkSparseImageMemoryBindInfo *bind)
+radv_sparse_image_bind_memory(struct radv_device *device, const VkSparseImageMemoryBindInfo *bind, uint32_t wait_count,
+                              const struct vk_sync_wait *waits)
 {
    VK_FROM_HANDLE(radv_image, image, bind->image);
    const struct radv_physical_device *pdev = radv_device_physical(device);
@@ -158,7 +161,7 @@ radv_sparse_image_bind_memory(struct radv_device *device, const VkSparseImageMem
       if (whole_subres) {
          uint64_t size = (uint64_t)aligned_extent_width * aligned_extent_height * aligned_extent_depth * bs;
          result = radv_bo_virtual_bind(device, &image->vk.base, image->bindings[0].bo, offset, size,
-                                       mem ? mem->bo : NULL, mem_offset);
+                                       mem ? mem->bo : NULL, mem_offset, wait_count, waits);
          if (result != VK_SUCCESS)
             return result;
       } else {
@@ -173,7 +176,8 @@ radv_sparse_image_bind_memory(struct radv_device *device, const VkSparseImageMem
 
                result = radv_bo_virtual_bind(device, &image->vk.base, image->bindings[0].bo, bo_offset, size,
                                              mem ? mem->bo : NULL,
-                                             mem_offset + (uint64_t)mem_y_increment * y + mem_z_increment * z);
+                                             mem_offset + (uint64_t)mem_y_increment * y + mem_z_increment * z,
+                                             wait_count, waits);
                if (result != VK_SUCCESS)
                   return result;
             }
@@ -188,19 +192,22 @@ static VkResult
 radv_queue_submit_bind_sparse_memory(struct radv_device *device, struct vk_queue_submit *submission)
 {
    for (uint32_t i = 0; i < submission->buffer_bind_count; ++i) {
-      VkResult result = radv_sparse_buffer_bind_memory(device, submission->buffer_binds + i);
+      VkResult result = radv_sparse_buffer_bind_memory(device, submission->buffer_binds + i,
+                                                       submission->wait_count, submission->waits);
       if (result != VK_SUCCESS)
          return result;
    }
 
    for (uint32_t i = 0; i < submission->image_opaque_bind_count; ++i) {
-      VkResult result = radv_sparse_image_opaque_bind_memory(device, submission->image_opaque_binds + i);
+      VkResult result = radv_sparse_image_opaque_bind_memory(device, submission->image_opaque_binds + i,
+                                                             submission->wait_count, submission->waits);
       if (result != VK_SUCCESS)
          return result;
    }
 
    for (uint32_t i = 0; i < submission->image_bind_count; ++i) {
-      VkResult result = radv_sparse_image_bind_memory(device, submission->image_binds + i);
+      VkResult result = radv_sparse_image_bind_memory(device, submission->image_binds + i,
+                                                      submission->wait_count, submission->waits);
       if (result != VK_SUCCESS)
          return result;
    }
